lib/core/batchingReceiver.ts

```diff
@@ -79,35 +79,40 @@ export class BatchingReceiver extends MessageReceiver {
       let onReceiveMessage: OnAmqpEventAsPromise;
       let onSessionClose: OnAmqpEventAsPromise;
       let onReceiveClose: OnAmqpEventAsPromise;
+      let onReceiveDrain: OnAmqpEvent;
       let onReceiveError: OnAmqpEvent;
       let onSessionError: OnAmqpEvent;
       let waitTimer: any;
       let maxMessageWaitTimer: any;
-      const resetCreditWindow = () => {
-        this._receiver!.setCreditWindow(0);
-        this._receiver!.addCredit(0);
-      };
+
       // Final action to be performed after maxMessageCount is reached or the maxWaitTime is over.
       const finalAction = () => {
         if (maxMessageWaitTimer) {
           clearTimeout(maxMessageWaitTimer);
         }
-        // Resetting the mode. Now anyone can call start() or receive() again.
+
+        clearTimeout(waitTimer);
+
         if (this._receiver) {
           this._receiver.removeListener(ReceiverEvents.receiverError, onReceiveError);
           this._receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
           this._receiver.session.removeListener(SessionEvents.sessionError, onSessionError);
-        }
 
-        clearTimeout(waitTimer);
-        resetCreditWindow();
-        this.isReceivingMessages = false;
-        log.batching(
-          "[%s] Resolving the promise with received list of messages: %O.",
-          this._context.namespace.connectionId,
-          brokeredMessages
-        );
-        resolve(brokeredMessages);
+          if (this._receiver.credit > 0) {
+            log.batching("[%s] Draining leftover credits (%s).", this._context.namespace.connectionId, this._receiver.credit);
+
+            // Setting drain must be accompanied by a flow call (aliased to addCredit in this case).
+            this._receiver.drain = true;
+            this._receiver.addCredit(1);
+          } else {
+            this._receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
+            this.isReceivingMessages = false;
+            log.batching("[%s] Resolving the promise with received list of messages: %O.", this._context.namespace.connectionId, brokeredMessages);
+            resolve(brokeredMessages);
+          }
+        } else {
+          this.isReceivingMessages = false;
```

Comment:
Why not resolve the messages at this point?

Reply:
Resolving here would allow the caller to begin a new batch immediately, where the batchingReciever isn't actually ready for a new batch until the receiver_drained event is received.

Reply:
> Resolving here would allow the caller to begin a new batch immediately, where the batchingReciever isn't actually ready for a new batch until the receiver_drained event is received.

That applies to the `if (this._receiver.credit > 0)` block.

I am referring to the `else` block at line 114. This code path is reached if the receiver is removed from cache i.e `this.receiver` is `null` or `undefined`. At this point there is no way for us to know if there are pending credits or not.

Reply:
Ah OK, I see. That was a miss on my part, and my manual testing didn't hit that case. I have updated the branch.